torch/csrc/dynamo/python_compiled_autograd.cpp

```diff
@@ -0,0 +1,487 @@
+          compiler_call.node_calls.lookup(fn).needed = false;
+        }
+      }
+      auto it = dependencies.find(edge.function.get());
+      TORCH_CHECK(it != dependencies.end());
+      if (--it->second == 0) {
+        dependencies.erase(it);
+        worklist.emplace_back(edge.function);
+      }
+    }
+  }
+
+  // TODO(jansel): some dynamic sizes seem to be ints not symints
+  if (!cache->check_dynamic_sizes(compiler_call)) {
+    // cache miss, need to capture FX graph
+    THPObjectPtr py_compiler(
+        check(PyObject_CallNoArgs((the_autograd_compiler))));
+    TraceState state = call_begin_capture(
+        py_compiler, *cache, compiler_call, output_edges.size());
+    InputBuffers input_buffers;
+
+    for (NodeCall* call_ptr : calls) {
+      NodeCall& call = *call_ptr;
+      // TODO(jansel): consider adding some of this stuff:
+      // guard(local_graph_task); NodeGuard ndguard(task.fn_); const auto
+      // opt_parent_stream = (*func).stream(c10::DeviceType::CUDA);
+      // c10::OptionalStreamGuard parent_stream_guard{opt_parent_stream};
+      // CheckpointValidGuard cpvguard(graph_task);
+      // at::getStepCallbacksUnlessEmpty(at::RecordScope::BACKWARD_FUNCTION);
+      // if (C10_UNLIKELY(step_callbacks.has_value())) { ... }
+
+      variable_list inputs = input_buffers.lookup(call.node.get()).buffer;
+
+      if (call.tensor_pre_hooks.size() + call.pre_hooks.size() > 0) {
+        THPObjectPtr pyinputs(THPVariable_WrapList(inputs));
+        for (const auto& hook : call.tensor_pre_hooks) {
+          pyinputs = check(PyObject_CallMethod(
+              py_compiler,
+              "tensor_pre_hook",
+              "Oii",
+              pyinputs.get(),
+              hook.first,
+              hook.second));
+        }
+        for (const auto hook : call.pre_hooks) {
+          pyinputs = check(PyObject_CallMethod(
+              py_compiler.get(), "pre_hook", "Oi", pyinputs.get(), hook));
+        }
+        inputs = THPVariable_UnpackList(pyinputs);
+      }
+      for (const auto& graph_output : call.graph_output) {
+        int input_nr = graph_output.first;
+        int output_index = graph_output.second;
+        TORCH_CHECK(output_index < static_cast<int>(state.outputs.size()));
+        TORCH_CHECK(!state.outputs[output_index].defined());
+        state.outputs[output_index] = inputs[input_nr];
+      }
+      if (!call.needed) {
+        continue;
+      }
+
+      SwapSavedVariables saved(state, call.node);
+      variable_list outputs = call.node->apply_with_saved(inputs, saved);
+      saved.before(call.node->next_edges());
+      validate_outputs(
+          call.node->next_edges(), outputs, [&](const std::string& msg) {
+            std::ostringstream ss;
+            ss << "[Compiled Autograd Tracing: " << call.node->name() << "] "
+               << msg;
+            return ss.str();
+          });
+      saved.after(call.node->next_edges());
+
+      if (call.post_hooks.size() > 0 && call.needed) {
```

Comment:
since we already continue above, do we need this call.needed?

Reply:
Removed